### DISCUSSION TITLE
Join query on 2 tables based not on id

### DISCUSSION THREAD

**Discussion by Pitu** [prisma/prisma#6024] [Q&A] [answered]
I have 2 tables called `packs` and `stickers`. 
A pack can have many stickers, and stickers can only be part of one pack. Now the tricky part is that the reference between those 2 tables has a different name. `packs.lineId` has the same value as `stickers.packId`.
Given the schema below, I'm not sure how to make the join query work.
```prisma
generator client {
	provider = "prisma-client-js"
}

datasource db {
	provider = "sqlite"
	url = "file:/database.sqlite"
}

model packs {
	id         Int        @id @default(autoincrement())
	lineId     Int?	      @unique
	name       String?
	animated   Boolean?
	count      Int?
	subscribed Int?       @default(0)
	enabled    Boolean?   @default(true)
	stickers   stickers[]
}

model stickers {
	id     Int     @id @default(autoincrement())
	pack   packs   @relation(fields: [packId], references: [lineId])
	packId Int
	lineId Int
	file   String?
}
```
And this is what my query looks like right now
```js
const packs = await prisma.packs.findMany({
	select: {
		id: true,
		name: true,
		count: true,
		lineId: true,
		animated: true,
		stickers: {
			select: {
				file: true,
				lineId: true
			}
		}
	}
});
```
What's wrong here is that `stickers[]` is coming up as an empty array since it's trying to match it to `packs.id` instead of `packs.lineId` and I'm not quite sure how to fix this.
Any help would be greatly appreciated!

**Comment by ryands17** [comment, marked as answer]
@Pitu 👋 

You don't need a `lineId` in your `stickers` model, just having it in the `packs` model should work:

```prisma
model packs {
  id         Int        @id @default(autoincrement())
  lineId     Int?       @unique
  name       String?
  animated   Boolean?
  count      Int?
  subscribed Int?       @default(0)
  enabled    Boolean?   @default(true)
  stickers   stickers[]
}

model stickers {
  id     Int     @id @default(autoincrement())
  pack   packs?  @relation(fields: [packId], references: [lineId])
  packId Int?
  file   String?
}
```

I ran the following query:

```ts
  await prisma.packs.create({
    data: {
      lineId: 1,
      stickers: { create: [{ file: 'file 1' }, { file: 'file 2' }] },
    },
  })

  const packs = await prisma.packs.findMany({
    select: {
      id: true,
      name: true,
      count: true,
      lineId: true,
      animated: true,
      stickers: {
        select: {
          file: true,
        },
      },
    },
})
```

And it gives me the correct result:

![image](https://user-images.githubusercontent.com/19697099/110456216-4c440f80-80ef-11eb-8d1c-5ad175cb7014.png)

**Reply by Pitu** [reply]
Thank you for your reply @ryands17

This doesn't seem to be working for me. The reason why `lineId` is needed in both `packs` and `stickers` is because each entity has it's own `lineId`, and tha's also the reason why `stickers.packId` is a reference to `packs.lineId`. I'll try to illustrate with what I'm seeing in DBeaver at the moment with my current populated database:

Given this entry in `packs` with `lineId` with a value of `14418`, fetching this pack with the above query should be returning the stickers depicted below:
![image](https://user-images.githubusercontent.com/7425261/110460792-02205580-8112-11eb-83a0-366f73c44a24.png)
![image](https://user-images.githubusercontent.com/7425261/110460945-2e3bd680-8112-11eb-8d2d-693af9d05796.png)

Yet when running the query mentioned above my output is the following: 
```json
{
  "packs": {
    "id": 318,
    "name": "BROWN&FRIENDS Large Letters",
    "count": 24,
    "lineId": 14418,
    "animated": true,
    "stickers": []
  }
}
```
This database was already in production and I ran the migrations necessary so that it matches the schema I posted above, yet it's not behaving like I intended and I can't figure out why.

**Reply by ryands17** [reply]
The `lineId` in the `Stickers` table is not needed at all as the reference is just from the `lineId` in the `packs` model to the `packId` in the `stickers` model.

If you could share some dummy data then I can seed it and try it on the model right now. Else the model looks correct and I can easily get the references. As per your example, the `packs` model item where the `lineId` is **14,418** should fetch all the `stickers` having `packId` **14,418**. So the `lineId` in the `stickers` model will never be used.

**Reply by Pitu** [reply]
The `lineId` in the `stickers` model is used by the application consuming the API but not for the reference as you point out, that's why it needs to be there and can't be removed.

As for dummy data, let me know if this helps!
```sql
INSERT INTO packs (id, lineId, name, animated, count, subscribed, enabled) VALUES(318, 14418, 'BROWN&FRIENDS Large Letters', 1, 24, 0, 1);
INSERT INTO packs (id, lineId, name, animated, count, subscribed, enabled) VALUES(630, 4409, 'Animated EVANGELION', 1, 24, 0, 1);
```
```sql
INSERT INTO stickers (id, packId, lineId, file) VALUES(11105, 14418, 183623494, '183623494.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11106, 14418, 183623495, '183623495.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11107, 14418, 183623496, '183623496.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11108, 14418, 183623497, '183623497.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11109, 14418, 183623498, '183623498.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11110, 14418, 183623499, '183623499.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11111, 14418, 183623500, '183623500.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11112, 14418, 183623501, '183623501.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11113, 14418, 183623502, '183623502.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11114, 14418, 183623503, '183623503.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11115, 14418, 183623504, '183623504.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11116, 14418, 183623505, '183623505.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11117, 14418, 183623506, '183623506.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11118, 14418, 183623507, '183623507.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11119, 14418, 183623508, '183623508.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11120, 14418, 183623509, '183623509.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11121, 14418, 183623510, '183623510.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11122, 14418, 183623511, '183623511.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11123, 14418, 183623512, '183623512.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11124, 14418, 183623513, '183623513.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11125, 14418, 183623514, '183623514.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11126, 14418, 183623515, '183623515.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11127, 14418, 183623516, '183623516.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(11128, 14418, 183623517, '183623517.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22241, 4409, 5437028, '5437028.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22242, 4409, 5437029, '5437029.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22243, 4409, 5437030, '5437030.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22244, 4409, 5437031, '5437031.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22245, 4409, 5437032, '5437032.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22246, 4409, 5437033, '5437033.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22247, 4409, 5437034, '5437034.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22248, 4409, 5437035, '5437035.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22249, 4409, 5437036, '5437036.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22250, 4409, 5437037, '5437037.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22251, 4409, 5437038, '5437038.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22252, 4409, 5437039, '5437039.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22253, 4409, 5437040, '5437040.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22254, 4409, 5437041, '5437041.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22255, 4409, 5437042, '5437042.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22256, 4409, 5437043, '5437043.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22257, 4409, 5437044, '5437044.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22258, 4409, 5437045, '5437045.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22259, 4409, 5437046, '5437046.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22260, 4409, 5437047, '5437047.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22261, 4409, 5437048, '5437048.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22262, 4409, 5437049, '5437049.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22263, 4409, 5437050, '5437050.gif');
INSERT INTO stickers (id, packId, lineId, file) VALUES(22264, 4409, 5437051, '5437051.gif');
```

**Reply by ryands17** [reply]
Seeded and ran this query:

```ts
let data = await prisma.packs.findMany({ include: { stickers: true } })
```

And the relations still work fine:

![image](https://user-images.githubusercontent.com/19697099/110464847-a9dd5980-80f9-11eb-8eb4-a1c35a72e104.png)

If this an already existing database, are the relations set properly in the database as well? I have tried on a clean database with Prisma version `2.18.0` and the above seed data. Could you try on a clean DB and check? If that works, seems like a foreign key issue.

**Reply by Pitu** [reply]
It was the database indeed!
I exported all the data and imported it on a clean database and the output is exactly what I was trying to get this entire time.
What's weird is that I used Prisma Migrate to alter the database based on the migration files it generated, but I guess something went wrong along the way even though the commands were all successful. 

Thank you so much for you help!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
